packages/captp/src/captp.js

```diff
@@ -169,6 +198,21 @@ export function makeCapTP(ourId, rawSend, bootstrapObj = undefined, opts = {}) {
     // eslint-disable-next-line no-use-before-define
     const pr = makeRemoteKit(questionID);
     questions.set(questionID, pr);
+
+    // To fix #2846:
+    // We return 'p' to the handler, and the eventual resolution of 'p' will
+    // be used to resolve the caller's Promise, but the caller never sees 'p'
+    // itself. The caller got back their Promise before the handler ever got
+    // invoked, and thus before queueMessage was called. If that caller
+    // passes the Promise they received as argument or return value, we want
+    // it to serialize as resultVPID. And if someone passes resultVPID to
+    // them, we want the user-level code to get back that Promise, not 'p'.
+    lastPromiseID += 1;
+    const promiseID = lastPromiseID;
+    const resultVPID = `p+${promiseID}`;
+    valToSlot.set(pr.p, resultVPID);
+    slotToVal.set(resultVPID, pr.p);
```

Comment:
If I remain tardy in reviewing this PR (sorry), would it be useful to extract this as a distinct PR so it can go in asap?

Reply:
We agreed that the fix doesn't need to go in immediately, since it's really only for deploy scripts.